### PR TITLE
JBTM-3286 Expose OpenAPI and Swagger definitions for LRA coordinator

### PR DESCRIPTION
**Pull request by xstefank**
https://issues.redhat.com/browse/JBTM-3286

!MAIN !TOMCAT !AS_TESTS !RTS !JACOCO !XTS !QA_JTA !QA_JTS_JACORB !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !BLACKTIE !PERF LRA !NO_WIN !DB_TESTS !mysql !db2 !postgres !oracle

### CODE REVIEW

**Comment by xstefank** [comment]
The test failure is tracked in different JBTM. It isn't caused by this change. @mmusgrov @tomjenkinson @ochaloup would it be possible to set up `-fae` flag for LRA profile? so we can get the basic tests and the rest of the TCK running even on failures?

**Comment by mmusgrov** [comment]
@xstefank I approved it but why didn't you add the dependency to the lra-coordinator-jar module since that is where we define the openapi annotations?

**Comment by mmusgrov** [comment]
We define 3 modules to support various containers:

1 lra-coordinator - quarkus
2. lra-coordinator-thorntail - thorntail
3. lra-coordinator-jar - any JAX-RS container such as wildfly

**Comment by xstefank** [comment]
@mmusgrov the dependency is Quarkus specific. I will take a look into thorntail also. But we can't add these runtime specific dependencies to the lra-coordinator-jar as they would be pulled in for instance wildfly you mentioned.

**Comment by xstefank** [comment]
@mmusgrov I added openapi and swagger to thorntail coordinator. However, I found out the thorntail coordinator can't run - https://issues.redhat.com/browse/JBTM-3287 and swagger isn't working (openapi works as expected). Is swagger something necessary for thorntail coordinator? If so I will file it as a new JIRA.

**Comment by mmusgrov** [comment]
> @mmusgrov I added openapi and swagger to thorntail coordinator. However, I found out the thorntail coordinator can't run - https://issues.redhat.com/browse/JBTM-3287 and swagger isn't working (openapi works as expected). Is swagger something necessary for thorntail coordinator? If so I will file it as a new JIRA.

Could we go back to using swagger annotations since those seemed to work with any JAX-RS container? Unless those don't work with quarkus in which case we'd need to investigate getting it working with WildFly. Thorntail does not matter so much since it will be end of life soon.

**Comment by xstefank** [comment]
@ochaloup the dependencies have been added to this PR in force push already. Also please do the narayana.sh script change in a separate PR to follow the project guidelines.

@mmusgrov @ochaloup about swagger. When I was moving coordinator to Quarkus I had to change the tag defining annotations from swagger to MP OpenAPI. Quarkus is able to pick these annotations to also expose the swagger interface (directly in the application). Thorntail would need to have swagger annotations (also it does not package with swagger directly, only exposes swagger.json). So with what we have now (MP OpenAPI annotations), we can have both OpenAPI and Swagger for quarkus based coordinator and OpenAPI also for thorntail coordinator but if you would want to have swagger (JSON) in thorntail coordinator then we would need to add swagger dependency to lra-coordinator-jar module and duplicate every MP OpenAPI annotation with its swagger counterpart (with the same string, parameters) which will be messy in my opinion. I would just drop the requirement for swagger in thorntail coordinator.

I read also now that swagger can import OpenAPI documents - https://github.com/swagger-api/swagger-editor/blob/master/docs/import.md.

OpenAPI will definitely work in WildFly or in any other container that support MP OpenAPI.

So I would drop the requirement for swagger directly in the thorntail coordinator. I would expect that we switch to quarkus as a main coordinator soon in everything.

**Comment by mmusgrov** [comment]
> @ochaloup the dependencies have been added to this PR in force push already. Also please do the narayana.sh script change in a separate PR to follow the project guidelines.

If the PR requires a change to narayana.sh then do it in this PR along with your other changes. There is no policy that requires you to raise 2 PR's that address the same fix (in fact to do otherwise would obfuscate the tracking of bug fixes).

>
> @mmusgrov @ochaloup about swagger. When I was moving coordinator to Quarkus I had to change the tag defining annotations from swagger to MP OpenAPI. Quarkus is able to pick these annotations to also expose the swagger interface
(directly in the application). Thorntail would need to have swagger annotations (also it does not package with swagger directly, only exposes swagger.json). So with what we have now (MP OpenAPI annotations), we can have both OpenAPI and Swagger for quarkus based coordinator and OpenAPI also for thorntail coordinator but if you would want to have swagger (JSON) in thorntail coordinator then we would need to add swagger dependency to lra-coordinator-jar module and duplicate every MP OpenAPI annotation with its swagger counterpart (with the same string, parameters) which will be messy in my opinion. I would just drop the requirement for swagger in thorntail coordinator.
>
> I read also now that swagger can import OpenAPI documents - https://github.com/swagger-api/swagger-editor/blob/master/docs/import.md.
>
> OpenAPI will definitely work in WildFly or in any other container that support MP OpenAPI.
>
> So I would drop the requirement for swagger directly in the thorntail coordinator. I would expect that we switch to quarkus as a main coordinator soon in everything.

If it works with WildFly and Quarkus then that is good enough (since Thorntail is nearing its EOL - the 4.x stream is the last).

**Comment by xstefank** [comment]
@mmusgrov @ochaloup already raised #1599 which covers -fae. This change is not targetted for this issue. It was just a general idea so I believe we should be good on this front.

I removed the swagger dependency from thorntail coordinator so now we have OpenAPI in both quarkus and thorntail and Quarkus also have swagger-ui on top of it built-in. This PR should be good to be merged now.

**Comment by mmusgrov** [comment]
> @mmusgrov @ochaloup already raised #1599 which covers -fae. This change is not targetted for this issue. It was just a general idea so I believe we should be good on this front.

Okay

> I removed the swagger dependency from thorntail coordinator so now we have OpenAPI in both quarkus and thorntail and Quarkus also have swagger-ui on top of it built-in. This PR should be good to be merged now.

So do the OpenAPI annotations work for WildFly too?

**Comment by xstefank** [comment]
@mmusgrov they should work. However, I am not able to deploy lra-coordinator-jar to Wildfly. Should I be able to so? If yes, then we need a new JBTM for this problem.

**Comment by mmusgrov** [comment]
> @mmusgrov they should work. However, I am not able to deploy lra-coordinator-jar to Wildfly. Should I be able to so? If yes, then we need a new JBTM for this problem.

We have a [maven profile](https://github.com/jbosstm/narayana/blob/master/rts/lra/lra-coordinator-jar/pom.xml#L156) that runs LRA recovery tests on wildfly (using wildfly-arquillian-container-managed artefact).

**Comment by xstefank** [comment]
If it runs in arquillian it doesn't mean it must deploy on wildfly. I tried with the latest wildfly master and lra-coordinator.jar from lra-coordinator-jar module doesn't deploy. Let me know if I should look into it.

**Comment by xstefank** [comment]
@ochaloup we can probably change it to mp-openapi API dependency. Let me do that.

**Comment by xstefank** [comment]
@ochaloup changed to MP openapi.

**Comment by mmusgrov** [comment]
> If it runs in arquillian it doesn't mean it must deploy on wildfly. 

I'm not sure what you mean but we run our arq profile as part of our standard testing.
If you look at the link I referenced from my comment you will see that we run the tests against the latest WildFly.

> I tried with the latest wildfly master and lra-coordinator.jar from lra-coordinator-jar module doesn't deploy. Let me know if I should look into it.

What problem did you hit?

**Comment by xstefank** [comment]
@mmusgrov in arquillian you are building custom deployment using the shrinkwrap. This is is a different archive than what is produced by the lra-coordinator-jar module (it's actually a war). So if lra-coordinator-jar should be deployable on its own then we need a new JBTM and this PR can be merged I think.

**Comment by mmusgrov** [comment]
> @mmusgrov in arquillian you are building custom deployment using the shrinkwrap. This is is a different archive than what is produced by the lra-coordinator-jar module (it's actually a war). So if lra-coordinator-jar should be deployable on its own then we need a new JBTM and this PR can be merged I think.

So does that mean arquilian testing is insufficient?

What problem did you hit?

**Comment by xstefank** [comment]
@mmusgrov 
```
Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type LRAService with qualifiers @Default
  at injection point [BackedAnnotatedField] @Inject io.narayana.lra.coordinator.api.RecoveryCoordinator.lraService
  at io.narayana.lra.coordinator.api.RecoveryCoordinator.lraService(RecoveryCoordinator.java:0)
WELD-001474: Class io.narayana.lra.coordinator.domain.service.LRAService is on the classpath, but was ignored because a class it references was not found: com.arjuna.ats.arjuna.recovery.RecoveryModule from [Module "deployment.lra-coordinator.jar" from Service Module Loader].
```

like I said if this is required then I'll open a new JBTM. Let's not continue this discussion on this PR, please.

**Comment by mmusgrov** [comment]
> @mmusgrov
> 
> ```
> Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type LRAService with qualifiers @Default
>   at injection point [BackedAnnotatedField] @Inject io.narayana.lra.coordinator.api.RecoveryCoordinator.lraService
>   at io.narayana.lra.coordinator.api.RecoveryCoordinator.lraService(RecoveryCoordinator.java:0)
> WELD-001474: Class io.narayana.lra.coordinator.domain.service.LRAService is on the classpath, but was ignored because a class it references was not found: com.arjuna.ats.arjuna.recovery.RecoveryModule from [Module "deployment.lra-coordinator.jar" from Service Module Loader].
> ```
> 
> like I said if this is required then I'll open a new JBTM. Let's not continue this discussion on this PR, please.

Okay go ahead and raise a JIRA for.

**Comment by mmusgrov** [comment]
> > @mmusgrov
> > ```
> > Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type LRAService with qualifiers @Default
> >   at injection point [BackedAnnotatedField] @Inject io.narayana.lra.coordinator.api.RecoveryCoordinator.lraService
> >   at io.narayana.lra.coordinator.api.RecoveryCoordinator.lraService(RecoveryCoordinator.java:0)
> > WELD-001474: Class io.narayana.lra.coordinator.domain.service.LRAService is on the classpath, but was ignored because a class it references was not found: com.arjuna.ats.arjuna.recovery.RecoveryModule from [Module "deployment.lra-coordinator.jar" from Service Module Loader].
> > ```
> > 
> > 
> > like I said if this is required then I'll open a new JBTM. Let's not continue this discussion on this PR, please.
> 
> Okay go ahead and raise a JIRA for.

@xstefank BTW we are in the process of producing a wildfly galleon pack for the coordinator. So we don't need a new JIRA.

File: rts/lra/lra-coordinator-jar/pom.xml [hunk 1]
```diff
@@ -81,9 +81,9 @@
             <scope>test</scope>
         </dependency>
         <dependency>
-            <groupId>io.smallrye</groupId>
-            <artifactId>smallrye-open-api</artifactId>
-            <version>1.1.19</version>
+            <groupId>org.eclipse.microprofile.openapi</groupId>
+            <artifactId>microprofile-openapi-api</artifactId>
+            <version>1.1.2</version>
```
Comment (ochaloup) [inline rts/lra/lra-coordinator-jar/pom.xml:86]: @xstefank just to understand, should this really be in `compile` scope? as this is expected to be deployed to WildFly that should be rather the WFLY module dependency which will be used not the one compiled in, right?
and that side note, the version should be probably configurable?

Comment (xstefank) [reply rts/lra/lra-coordinator-jar/pom.xml]: it can. Sorry, I forgot about this comment.

**Pull request merged**